source/custom/nisync_service.custom.cpp

```diff
@@ -26,14 +26,23 @@ ::grpc::Status NiSyncService::ReadMultipleTriggerTimeStamp(::grpc::ServerContext
     ViUInt32 timestamps_read {};
     auto status = library_->ReadMultipleTriggerTimeStamp(vi, terminal, timestamps_to_read, timeout, time_seconds_buffer, time_nanoseconds_buffer, time_fractional_nanoseconds_buffer, detected_edge_buffer, &timestamps_read);
     response->set_status(status);
-    if (status == VI_SUCCESS || status == NISYNC_ERROR_DRIVER_TIMEOUT) {
-      response->set_timestamps_read(timestamps_read);
+    if (status < VI_SUCCESS) {
```

Comment:
Hmm, this one's weird. It seems in https://github.com/ni/grpc-device/pull/209 this was added to still set the timestamps_read field even when the error is `NISYNC_ERROR_DRIVER_TIMEOUT`. I'm wondering if we should keep that behavior by checking `status < VI_SUCCESS && status != NISYNC_ERROR_DRIVER_TIMEOUT`.

Reply:
FYI, DAQmx read functions can also return (partial) samples along with a timeout error:

https://www.ni.com/docs/en-US/bundle/ni-daqmx-c-api-ref/page/daqmxcfunc/daqmxreadanalogf64.html

> A value of 0 indicates to try once to read the requested samples. If all the requested samples are read, the function is successful. Otherwise, the function returns a timeout error and returns the samples that were actually read.

Reply:
That's going to be fun.

Reply:
I guess that's probably a bug.

Would it work to incorporate that into `status_ok`? That's implemented per-driver to decide what status codes are errors vs warnings (where "warnings" we copy outputs but "errors" we don't).

So basically the question is "is it valid to treat a timeout error as a warning throughout the daq/sync API?"

Or are there other entrypoints (potentially) where timeouts need to be handled as errors? If so, we could add an optional `errors_to_handle_as_warnings: List[int]` to function metadata.

Reply:
@astarche that (adding the check to `status_ok`) sounds right to me except that it's usually an inline method in the generated file instead of being a member method that I could implement in the custom file. In the case of this driver (nisync) the generated check does not cover this same error code. How should we pull that off?

Reply:
First: for the sync method in this diff. This is `custom` code, so I think it makes sense to take @reckenro 's suggestion: restore the original condition and avoid a regression here.

I'm thinking more of the issue with DAQ that Brad pointed out and maybe a general fix.

The `status_ok` methods codegen from an expression in config.py like [this one for DAQ](https://github.com/ni/grpc-device/blob/cca4d8a11a1cbec001c6267137c3e5e2b2d0e9ec/source/codegen/metadata/nidaqmx/config.py#L135). So, if the expression can be changed to something like `'status_ok': 'status >= 0 || status == DAQmx_Timeout', then that's an easy fix.

Otherwise, we'd need to change [the codegen for the callers of `status_ok` ](https://github.com/ni/grpc-device/blob/cca4d8a11a1cbec001c6267137c3e5e2b2d0e9ec/source/codegen/templates/service_helpers.mako#L730) so they can conditionally check additional error codes based on some new field in the `function_data`.

Reply:
> So basically the question is "is it valid to treat a timeout error as a warning throughout the daq/sync API?"

No, I don't think this is valid. 

Not all timeout errors correspond to a user-specified timeout parameter. Some timeout errors indicate that a hardware operation didn't complete within a fixed time. For example, when writing to an EEPROM or flash ROM, the driver may poll the status register to determine when the write cycle is complete. It would be inappropriate to hide timeout errors from `DAQmxSelfCal` or `DAQmxCloseExtCal`.

If the driver has separate error codes for user-specified timeouts and hardware polling timeouts, you might be able to get by with it, but I don't think this is always the case.

Let's look at DAQmx's timeout error codes:
- DAQmxErrorSamplesNotYetAvailable: used for DAQmx read; can return acquired samples
- DAQmxErrorCannotReadRelativeToRefTrigUntilDone; used for DAQmx read; can't return acquired samples 
- DAQmxErrorSamplesCanNotYetBeWritten: used for DAQmx write
- DAQmxErrorTimeoutExceeded: used for hardware polling and DAQmxAutoConfigureCDAQSyncConnections, which has a user-specified timeout

> Or are there other entrypoints (potentially) where timeouts need to be handled as errors? If so, we could add an optional errors_to_handle_as_warnings: List[int] to function metadata.

I think this is less likely to have unintended consequences.

> I'm thinking more of the issue with DAQ that Brad pointed out and maybe a general fix.
> 
>The status_ok methods codegen from an expression in config.py like [this one for DAQ](https://github.com/ni/grpc-device/blob/cca4d8a11a1cbec001c6267137c3e5e2b2d0e9ec/source/codegen/metadata/nidaqmx/config.py#L135). So, if the expression can be changed to something like 'status_ok': 'status >= 0 || status == DAQmx_Timeout, then that's an easy fix.

I think fixing this for DAQmx requires more thought and should be handled separately.

Do you even need to fix this for DAQmx?  Do the other DAQmx APIs that use exceptions (.NET and Python) return samples in timeout cases?

Do they return any other outputs for read/write errors? nidaqmx-python adds the number of samples read/written to the exception object, and this is not specific to timeout errors. 

If it's important to support these use cases, perhaps it would be better to allow clients to opt out of exception-throwing for specific calls.

Reply:
> Do you even need to fix this for DAQmx? For comparison, do the DAQmx APIs that use exceptions (.NET and Python) return samples in timeout cases?

Good question. It looks to me like the python API [just throws for any error from read](https://github.com/ni/nidaqmx-python/blob/1f5586385d4da2cfa922834adc77a1f768aff259/nidaqmx/_task_modules/read_functions.py#L31). So that's a good argument for "do nothing".

Reply:
I fixed the one issue Ryan brought up above and will leave well enough alone for now for other cases. I'm going to mark this issue resolved.